crates/provider/src/provider/web3_signer.rs

```diff
@@ -0,0 +1,173 @@
+    ///
+    /// The `address` is used to set the `from` field in the transaction requests.
+    ///
+    /// The remote signer's address _must_ be the same as the signer address provided here.
+    pub fn new(provider: P, address: Address) -> Self {
+        Self { provider, address, _pd: std::marker::PhantomData }
+    }
+
+    /// Signs a transaction request and return the raw signed transaction in the form of [`Bytes`].
+    ///
+    /// The returned [`Bytes`] can be used to broadcast the transaction to the network using
+    /// [`Provider::send_raw_transaction`].
+    ///
+    /// Sets the `from` field to the provided `address`.
+    pub async fn sign_transaction(
+        &self,
+        mut tx: N::TransactionRequest,
+    ) -> alloy_signer::Result<Bytes> {
+        tx.set_from(self.address);
+        self.provider.sign_transaction(tx).await.map_err(alloy_signer::Error::other)
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl<P, N> TxSigner<Signature> for Web3Signer<P, N>
+where
+    P: Provider<N>,
+    N: Network,
+    N::TransactionRequest: TransactionBuilder7702 + TransactionBuilder4844,
+{
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    async fn sign_transaction(
+        &self,
+        tx: &mut dyn SignableTransaction<Signature>,
+    ) -> alloy_signer::Result<Signature> {
+        let mut request = N::TransactionRequest::default();
+
+        // Basics
+        request.set_kind(tx.kind());
+        request.set_nonce(tx.nonce());
+        request.set_input(tx.input().clone());
+        request.set_value(tx.value());
+
+        if let Some(chain_id) = tx.chain_id() {
+            request.set_chain_id(chain_id);
+        }
+
+        // Gas related fields
+        request.set_gas_limit(tx.gas_limit());
+        let max_fee_or_gas_price = tx.max_fee_per_gas(); // Returns `gasPrice` if not dynamic fee.
+        if tx.is_dynamic_fee() {
+            request.set_max_fee_per_gas(max_fee_or_gas_price);
+            if let Some(max_priority_fee) = tx.max_priority_fee_per_gas() {
+                request.set_max_priority_fee_per_gas(max_priority_fee);
+            }
+        } else {
+            request.set_gas_price(max_fee_or_gas_price);
+        }
+
+        if let Some(access_list) = tx.access_list() {
+            request.set_access_list(access_list.clone());
+        }
+
+        if let Some(sidecar) = tx.blob_sidecar() {
+            request.set_blob_sidecar(sidecar.clone());
+        }
+
+        if let Some(auth) = tx.authorization_list() {
+            request.set_authorization_list(auth.to_vec());
+        }
+
+        let raw = self.sign_transaction(request).await?;
+
+        let envelope =
+            TxEnvelope::decode_2718(&mut raw.as_ref()).map_err(alloy_signer::Error::other)?;
+
+        Ok(*envelope.signature())
```

Comment:
we should add a check here that the signed tx actually matches what we received as signable tx because otherwise we return a signature for a different tx object.

for this we need to ensure that encode_for_signing is the same for both